decrypt/decrypt.go

```diff
@@ -1,42 +1,213 @@
 package decrypt
 
 import (
+	"errors"
 	"flag"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
 
+	"github.com/NBISweden/sda-cli/helpers"
+	"github.com/elixir-oslo/crypt4gh/keys"
+	"github.com/elixir-oslo/crypt4gh/streaming"
 	log "github.com/sirupsen/logrus"
 )
 
 // Help text and command line flags.
 
+// Usage text that will be displayed as command line help text when using the
+// `help decrypt` command
 var Usage = `
-USAGE: %s decrypt -key <private-key-file> [file(s)]
+USAGE: %s decrypt (-createKey <name>) -key <private-key-file> [file(s)]
 
 Decrypt: Encrypts files from the Sensitive Data Archive (SDA) with the provided
-         private key.
+         private key. If the private key is encrypted, the password can be
+		 supplied in the DECRYPT_PASSWORD environment variable, or at the
+		 interactive password prompt.
 `
+
+// ArgHelp is the suffix text that will be displayed after the argument list in
+// the module help
 var ArgHelp = `
   [file(s)]
         all flagless arguments will be used as filenames for decryption.`
 
+// Args is a flagset that needs to be exported so that it can be written to the
+// main program help
 var Args = flag.NewFlagSet("decrypt", flag.ExitOnError)
 
-var privateKey = Args.String("key", "",
+var privateKeyFile = Args.String("key", "",
 	"Private key to use for decrypting files.")
 
-// Main decryption function
-func Decrypt(args []string) {
+// Decrypt takes a set of arguments, parses them, and attempts to eiher create a
+// crypt4gh key pair (if -keyName is set) or decrypt the given data files with
+// the given private key file (otherwise).
+func Decrypt(args []string) error {
 	err := Args.Parse(args[1:])
 	if err != nil {
-		log.Fatalf("Argument parsing failed, reason: %v", err)
+		return fmt.Errorf("argument parsing failed, reason: %v", err)
 	}
 
+	// format input and output files
 	// Args() returns the non-flag arguments, which we assume are filenames.
-	files := Args.Args()
+	// All filenames are read into a struct together with their output filenames
+	files := []helpers.EncryptionFileSet{}
+	for _, filename := range Args.Args() {
+
+		// Set directory for the output file
+		unencryptedFilename := strings.TrimSuffix(filename, ".c4gh")
+
+		files = append(files, helpers.EncryptionFileSet{Encrypted: filename, Unencrypted: unencryptedFilename})
+	}
 
 	// Check that we have a private key to decrypt with
-	if *privateKey == "" {
-		log.Fatal("A private key is required to decrypt data")
+	if *privateKeyFile == "" {
+		return errors.New("a private key is required to decrypt data")
+	}
+
+	var privateKey *[32]byte
+
+	// try reading private key without password
+	privateKey, err = readPrivateKey(*privateKeyFile, "")
+	if err != nil {
+
+		// if there was an error, try again with the password
+		password, err := getPassword("DECRYPT_PASSWORD")
```

Comment:
Better ENV name for the private key password might be in order

Reply:
YES! I couldn't come up with one, so I just used something descriptive that wasn't tied to SDA or BP... any ideas? or should I make a flag for the ENV?

Reply:
C4GH_PASSWORD perhaps

Reply:
sure! :)